apps/assistants/models.py

```diff
@@ -171,33 +173,64 @@ def compare_with_model(self, new: Self, exclude_fields: list[str], early_abort=F
+            self.get_related_experiments_queryset().exists()
+            or self.get_related_pipeline_node_queryset().exists()
+            or self.get_related_experiments_with_pipeline_queryset().exists()
+        ):
             return False
-
-        if self.get_related_pipeline_node_queryset().exists():
-            return False
-
-        super().archive()
         if self.is_working_version:
-            # TODO: should this delete the assistant from OpenAI?
+            for (
+                version
+            ) in self.versions.all():  # first perform all checks so assistants are not archived prior to return False
+                if (
+                    version.get_related_experiments_queryset().exists()
+                    or version.get_related_pipeline_node_queryset().exists()
+                    or self.get_related_experiments_with_pipeline_queryset().exists()
+                ):
+                    return False
+            for version in self.versions.all():
+                delete_openai_assistant_task.delay(version.id)
             self.versions.update(is_archived=True, audit_action=AuditAction.AUDIT)
-        else:
-            delete_openai_assistant_task.delay(self.id)
 
+        super().archive()
+        delete_openai_assistant_task.delay(self.id)
         return True
 
-    def get_related_experiments_queryset(self):
-        return self.experiment_set.filter(is_archived=False)
+    def get_related_experiments_queryset(self, assistant_ids: list = None):
+        if assistant_ids:
+            return Experiment.objects.filter(
+                Q(working_version_id=None) | Q(is_default_version=True),
+                assistant_id__in=assistant_ids,
+                is_archived=False,
+            )
 
-    def get_related_pipeline_node_queryset(self):
-        from apps.pipelines.models import Node
+        return self.experiment_set.filter(Q(working_version_id=None) | Q(is_default_version=True), is_archived=False)
 
+    def get_related_pipeline_node_queryset(self, assistant_ids: list = None):
+        assistant_ids = assistant_ids if assistant_ids else [str(self.id)]
         return Node.objects.filter(type="AssistantNode").filter(
-            params__assistant_id=str(self.id),
+            pipeline__working_version_id=None,
+            params__assistant_id__in=assistant_ids,
             pipeline__is_archived=False,
         )
 
+    def get_related_experiments_with_pipeline_queryset(self, assistant_ids: list = None):
+        assistant_ids = assistant_ids if assistant_ids else [str(self.id)]
+        nodes = Node.objects.filter(type="AssistantNode").filter(
+            pipeline__working_version_id__isnull=False,
+            params__assistant_id__in=assistant_ids,
+            pipeline__is_archived=False,
+        )
+        if nodes.exists():
+            pipeline_ids = nodes.values_list("pipeline_id", flat=True)
+            return Experiment.objects.filter(
+                is_default_version=True,
```

Comment:
Do we care about working experiments as well?

Reply:
@stephherbers correct me if I'm wrong but I think we assume that a working version of an experiment must reference a working version of a pipeline which is covered in the query above.

Maybe a comment in code would help clarify the logic for posterity.